libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,560 @@
+			if (!zeroVisited)
+				_toVisit.emplace_front(_conditionalJump.zero);
+			if (!nonZeroVisited)
+				_toVisit.emplace_front(_conditionalJump.nonZero);
+			return nullopt;
+		},
+		[&](CFG::BasicBlock::FunctionReturn const& _functionReturn) -> std::optional<Stack>
+		{
+			yulAssert(_functionReturn.info, "");
+			Stack stack = _functionReturn.info->returnVariables | ranges::views::transform([](auto const& _varSlot){
+				return StackSlot{_varSlot};
+			}) | ranges::to<Stack>;
+			stack.emplace_back(FunctionReturnLabelSlot{});
+			return stack;
+		},
+		[&](CFG::BasicBlock::Terminated const&) -> std::optional<Stack>
+		{
+			return Stack{};
+		},
+	}, _block.exit);
+}
+
+void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const& _entry)
+{
+	list<CFG::BasicBlock const*> toVisit{&_entry};
+	set<CFG::BasicBlock const*> visited;
+
+	// TODO: check whether visiting only a subset of these in the outer iteration below is enough.
+	list<pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> backwardsJumps = collectBackwardsJumps(_entry);
+
+	while (!toVisit.empty())
+	{
+		// First calculate stack layouts without walking backwards jumps, i.e. assuming the current preliminary
+		// entry layout of the backwards jump target as the initial exit layout of the backwards-jumping block.
+		while (!toVisit.empty())
+		{
+			CFG::BasicBlock const *block = *toVisit.begin();
+			toVisit.pop_front();
+
+			if (visited.count(block))
+				continue;
+
+			if (std::optional<Stack> exitLayout = getExitLayoutOrStageDependencies(*block, visited, toVisit))
+			{
+				visited.emplace(block);
+				// We can skip the visit, if we have seen this precise exit layout already last time.
+				// Note: if the entire graph is revisited in the backwards jump check below, doing
+				//       this seems to break things; not sure why.
+				// Note: since I don't quite understand why doing this can break things, I comment
+				//       it out for now, since not aborting in those cases should always be safe.
+				// if (auto* previousInfo = util::valueOrNullptr(m_layout.blockInfos, block))
+				//	if (previousInfo->exitLayout == *exitLayout)
+				//		continue;
+				auto& info = m_layout.blockInfos[block];
+				info.exitLayout = *exitLayout;
+				info.entryLayout = propagateStackThroughBlock(info.exitLayout, *block);
+
+				for (auto entry: block->entries)
+					toVisit.emplace_back(entry);
+			}
+			else
+				continue;
+		}
+
+		// Determine which backwards jumps still require fixing and stage revisits of appropriate nodes.
+		for (auto [block, target]: backwardsJumps)
+			if (ranges::any_of(
+				m_layout.blockInfos[target].entryLayout,
+				[exitLayout = m_layout.blockInfos[block].exitLayout](StackSlot const& _slot) {
+					return !util::findOffset(exitLayout, _slot);
+				}
+			))
+			{
+				// This block jumps backwards, but does not provide all slots required by the jump target on exit.
+				// Therefore we need to visit the subgraph between ``target`` and ``block`` again.
```

Comment:
Why does the slot magically appear when we visit again?

Reply:
Example:
```
let i := 0
let j := 1
let k := 2
for {} lt(i, 4) {} {
  sstore(k,k)
}
sstore(i, j)
```

The first visit will start at the main exit and walk upwards to the loop condition saying it needs `i` and `j`. But that's not all I need for the loop condition, I also need `k`. So before visiting the loop condition, I also have to walk the path to the loop body. That visit move down up until the backwards jump - it doesn't yet know what the entry of the loop condition needs, so it just assumes it needs nothing. That way I get the `k` requirement: the loop condition exit is a conditional jump that combines the `[i, j]` and the `[k]` requirement. Then I can move through the loop condition and see what it needs on entry. That's the final layout - I won't ever need any more slots there.

But the loop body is not correct yet, since it started from ``[k]`` only. But in a second pass, I already know the loop condition entry, so instead of starting from an empty layout at the backwards jump, I directly start from `[i,j,k]` - the required slots have "magically" appeared :-).

Reply:
For nested loops it gets crazy then...

Reply:
Optimally, for nested loops, I'd *probably* fix the outer loop first, then the inner one... the algorithm doesn't take that into account and just fixes the backwards edges one by one, but eventually the layouts will stabilize anyways...